### PR TITLE
feat: embeddings similarity

### PR DESCRIPTION
**Pull request by daibhin**
## Problem

We want to use the embeddings of session replays being generated since https://github.com/PostHog/posthog/pull/20046 to see if we can extract similar recordings

## Changes

- Find embeddings for a target recording
- Use ClickHouse distance functions to find nearest neighbours
- Setup feature flag and simple frontend UI to test easily

## How did you test this code?

Behind a feature flag for now while we test

### CODE REVIEW

File: frontend/src/scenes/session-recordings/player/SimilarRecordings.tsx
Comment (pauldambra) [inline frontend/src/scenes/session-recordings/player/SimilarRecordings.tsx:1]: should we put a trigger button for it in here

Screenshot 2024-02-19 at 19 01 48

i can imagine just wanting a manual trigger for now while we test it out?

File: posthog/session_recordings/session_recording_api.py [hunk 1]
```diff
@@ -503,6 +504,35 @@ def summarize(self, request: request.Request, **kwargs):
...
+            raise exceptions.ValidationError("similar recordings is not enabled for this user")
+
+        recording = self.get_object()
+
+        if not SessionReplayEvents().exists(session_id=str(recording.session_id), team=self.team):
+            raise exceptions.NotFound("Recording not found")
+
+        recordings = similar_recordings(recording, self.team)
+        if recordings:
+            cache.set(cache_key, recordings, timeout=30)
```
Comment (pauldambra) [inline posthog/session_recordings/session_recording_api.py:530]: changed it so we don't cache empty recordings

**Review by pauldambra — CHANGES_REQUESTED** [review]
great start

File: frontend/src/scenes/session-recordings/player/SimilarRecordings.tsx
Comment (pauldambra) [reply frontend/src/scenes/session-recordings/player/SimilarRecordings.tsx:1]: feels like we're on the verge of inventing unsaved playlists / saved filters for clusterings or something...

so, we'd open an "unsaved" playlist that shows similar recordings to X, and then autosuggest title and description

title would be "recordings like {UUID}" description comes from chat gpt????

Comment (pauldambra) [reply frontend/src/scenes/session-recordings/player/SimilarRecordings.tsx:1]: but that can come later (or be totally different)

File: ee/session_recordings/ai/similar_recordings.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    return result[0][0] if len(result) > 0 else None
+
+
+def closest_embeddings(target: List[float], session_id: str, team_id: int):
+    query = """
+            SELECT
+                session_id,
+                -- distance function choice based on https://help.openai.com/en/articles/6824809-embeddings-frequently-asked-questions
+                -- OpenAI normalizes embeddings so L2 should produce the same score but is slightly slower
+                cosineDistance(embeddings, %(target)s) AS similarity_score
```
[the remark this answers is not in the document]
Comment (daibhin) [reply ee/session_recordings/ai/similar_recordings.py]: Found a good example on how to do that: https://clickhouse.com/blog/approximate-nearest-neighbour-ann-with-sql-powered-local-sensitive-hashing-lsh-random-projections#querying

File: frontend/src/scenes/session-recordings/player/SimilarRecordings.tsx
Comment (daibhin) [reply frontend/src/scenes/session-recordings/player/SimilarRecordings.tsx:1]: That certainly makes sense. Assuming the similarity measures work (which this PR will prove) that could well save us the problem of cluster naming. We don't have to be able to make the next leap and say _why_ they're similar, just that they're similar to some root recording

**Review by pauldambra — APPROVED** [review]
LGTM - really excited to see how this performs!

**Pull request merged**